File: src/Tests/Behavioral/GenericCompositeType/GenericCompositeType.go
package main

import "fmt"

// Option represents an optional value
type Option[T any] struct {
	value T
	valid bool
}

func NewOption[T any](value T) Option[T] {
	return Option[T]{
		value: value,
		valid: true,
	}
}

func NewEmptyOption[T any]() Option[T] {
	var zero T
	return Option[T]{
		value: zero,
		valid: false,
	}
}

func main() {
	// Pointer to a generic type (StarExpr)
	optionPtr := &Option[string]{
		value: "hello",
		valid: true,
	}
	fmt.Printf("Option value: %s, valid: %t\n", optionPtr.value, optionPtr.valid)

	// Slice of a generic type (ArrayType)
	options := []Option[int]{
		NewOption(42),
		NewEmptyOption[int](),
		NewOption(100),
	}
	fmt.Printf("Options count: %d\n", len(options))

	// Map with generic key and value types (MapType)
	cache := map[string]Option[float64]{
		"pi":  NewOption(3.14159),
		"e":   NewOption(2.71828),
		"phi": NewEmptyOption[float64](),
	}
	fmt.Printf("Cache size: %d\n", len(cache))

	// Channel of a generic type (ChanType)
	resultChan := make(chan Option[bool], 1)
	resultChan <- NewOption(true)
	result := <-resultChan
	fmt.Printf("Channel result valid: %t\n", result.valid)
}

File: src/Tests/Behavioral/GenericCompositeType/GenericCompositeType.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType] partial struct Option<T>
    where T : new()
{
    internal T value;
    internal bool valid;
}

public static Option<T> NewOption<T>(T value)
    where T : new()
{
    return new Option<T>(
        value: value,
        valid: true
    );
}

public static Option<T> NewEmptyOption<T>()
    where T : new()
{
    T zero = default!;
    return new Option<T>(
        value: zero,
        valid: false
    );
}

internal static void Main() {
    var optionPtr = Ꮡ(new Option<@string>(
        value: "hello"u8,
        valid: true
    ));
    fmt.Printf("Option value: %s, valid: %t\n"u8, (~optionPtr).value, (~optionPtr).valid);
    var options = new Option<nint>[]{
        NewOption<nint>(42),
        NewEmptyOption<nint>(),
        NewOption<nint>(100)
    }.slice();
    fmt.Printf("Options count: %d\n"u8, len(options));
    var cache = new map<@string, Option<float64>>{
        ["pi"u8] = NewOption<float64>(3.14159F),
        ["e"u8] = NewOption<float64>(2.71828F),
        ["phi"u8] = NewEmptyOption<float64>()
    };
    fmt.Printf("Cache size: %d\n"u8, len(cache));
    var resultChan = new channel<Option<bool>>(1);
    resultChan.ᐸꟷ(NewOption<bool>(true));
    var result = ᐸꟷ(resultChan);
    fmt.Printf("Channel result valid: %t\n"u8, result.valid);
}

} // end main_package
